x-pack/plugins/enterprise_search/public/applications/__mocks__/shallow_useeffect.mock.ts

```diff
@@ -4,14 +4,18 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
+let mockUmountHandler: () => void;
+
 jest.mock('react', () => ({
   ...(jest.requireActual('react') as object),
-  useEffect: jest.fn((fn) => fn()), // Calls on mount/every update - use mount for more complex behavior
+  useEffect: jest.fn((fn) => {
+    mockUmountHandler = fn();
```

Comment:
Sorry, I'm confused what this is accomplishing, can you give me more context?

Reply:
https://github.com/elastic/kibana/pull/88066/commits/2a20621f263c48b6cd77d1cab7eb77894d39f603

> Calling the unmount handler was calling useEffect again, which
can mess with expectations in specs that assume that useEffect will
only be called once.

If you have code like this:
```
useEffect(() => {
    doSomething();
    return () => cleanSomething();
  }, []);
```

When testing, you want to test 2 cases:
1. When the component mounts, `doSomething` is called.
2. When unmounting the component, `cleanSomething` is called.

Rendering with `shallow` will not trigger `useEffect`. To work around that, you can import this file, which will call your useEffect handler and allow you to assert that `doSomething` was called.


```
import '../../../__mocks__/shallow_useeffect.mock';

...

expect(doSomething).toHaveBeenCalled()
```

You also want to test that `cleanSomething ` is called after the component is unmounted. `shallow` is again no help here. This mock will give you direct access to the handler so you can call it yourself:

```
import { unmountHandler } from '../../../__mocks__/shallow_useeffect.mock';

...
unmountHandler();
expect(cleanSomething).toHaveBeenCalled()
```

The problem with the previous code, was that calling `unmountHandler` was calling `useEffect` AGAIN in order to provide access to the unmount handler.

```
import { unmountHandler } from '../../../__mocks__/shallow_useeffect.mock';

...
expect(doSomething).toHaveBeenCalledOnce() // true
unmountHandler();
expect(doSomething).toHaveBeenCalledOnce() // false
expect(cleanSomething).toHaveBeenCalled()
```

Instead of invoking `useEffect` again, I save off a reference to the unmount so it can be accessed later.

Hope that clarifies, it's a little difficult to explain.

Reply:
Thank you, this explanation makes sense

Reply:
@JasonStoltz it's not clear to me why you're checking `expect(doSomething).toHaveBeenCalledOnce()` multiple times in your example, isn't the important thing to check for after an unmount `cleanSomething()` and not anything else?

Reply:
The example is just to illustrate why `unmountHandler` was not accurately mocking how a component would actually work.

My use case was different.